pkg/extension/opampextension/opamp_agent_test.go

```diff
@@ -159,435 +49,482 @@ func TestApplyRemoteConfig(t *testing.T) {
-	assert.NoError(t, err)
-	assert.True(t, changed)
-	assert.NotEqual(t, len(o.effectiveConfig), 0)
-
-	cfg.AcceptsRemoteConfiguration = false
-	changed, err = o.applyRemoteConfig(rc)
-	assert.False(t, changed)
-	assert.Error(t, err)
-	assert.Equal(t, "OpAMP agent does not accept remote configuration", err.Error())
-}
-
-func TestShutdown(t *testing.T) {
-	cfg := createDefaultConfig().(*Config)
-	cfg.ClientConfig.Auth = nil
-	set := extensiontest.NewNopSettings()
-	o, err := newOpampAgent(cfg, set.Logger, set.BuildInfo, set.Resource)
-	assert.NoError(t, err)
-
-	// Shutdown with no OpAMP client
-	assert.NoError(t, o.Shutdown(context.Background()))
-}
-
-func TestStart(t *testing.T) {
-	d, err := os.MkdirTemp("", "opamp.d")
-	assert.NoError(t, err)
-	defer os.RemoveAll(d)
-
-	cfg := createDefaultConfig().(*Config)
-	cfg.ClientConfig.Auth = nil
-	cfg.RemoteConfigurationDirectory = d
-	set := extensiontest.NewNopSettings()
-	o, err := newOpampAgent(cfg, set.Logger, set.BuildInfo, set.Resource)
-	assert.NoError(t, err)
-
-	assert.NoError(t, o.Start(context.Background(), componenttest.NewNopHost()))
-}
-
-func TestReload(t *testing.T) {
-	d, err := os.MkdirTemp("", "opamp.d")
-	assert.NoError(t, err)
-	defer os.RemoveAll(d)
-
-	cfg := createDefaultConfig().(*Config)
-	cfg.ClientConfig.Auth = nil
-	cfg.RemoteConfigurationDirectory = d
-	set := extensiontest.NewNopSettings()
-	o, err := newOpampAgent(cfg, set.Logger, set.BuildInfo, set.Resource)
-	assert.NoError(t, err)
-
-	ctx := context.Background()
-	assert.NoError(t, o.Start(ctx, componenttest.NewNopHost()))
-	assert.NoError(t, o.Reload(ctx))
-}
-
-// To be removed when the OpAMP backend correctly advertises its URL.
-// This test is badly written, it reaches into unexported fields to test
-// their contents, but given that it will be removed in a few months, that
-// shouldn't matter too much from a big picture perspective.
-func TestHackSetEndpoint(t *testing.T) {
-	tests := []struct {
-		name         string
-		url          string
-		wantEndpoint string
-	}{
 		{
-			name:         "empty url defaults to config endpoint",
-			wantEndpoint: "wss://example.com",
+			name: "ApplyRemoteConfig",
+			setup: func() (*Config, extension.Settings) {
+				tempDir := t.TempDir()
+				cfg := createDefaultConfig().(*Config)
+				cfg.RemoteConfigurationDirectory = tempDir
+				set := extensiontest.NewNopSettings()
+
+				return cfg, set
+			},
+			validate: func(o *opampAgent, t *testing.T) {
+				path := filepath.Join("testdata", "opamp.d", "opamp-remote-config.yaml")
+				rc := remoteConfig(t, path)
+
+				changed, err := o.applyRemoteConfig(rc)
+				assert.NoError(t, err)
+				assert.True(t, changed)
+				assert.NotEqual(t, len(o.effectiveConfig), 0)
+
+				o.cfg.AcceptsRemoteConfiguration = false
+				changed, err = o.applyRemoteConfig(rc)
+				assert.False(t, changed)
+				assert.Error(t, err)
+				assert.Equal(t, "OpAMP agent does not accept remote configuration", err.Error())
```

Comment:
seems like there're more than 2 duplicates, should this be some type of function? 🤔